File: companion/app/src/main/kotlin/fr/lahocy/topolink/DistoBle.kt
package fr.lahocy.topolink

import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.content.Context
import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Leica DISTO Bluetooth Smart (D110 / D2 / D510 / D810…) en BLE.
 * Service 3ab10100-… ; distance 3ab10101-… (float32 LE, notification) ; commande 3ab10109-… ("g" = mesurer).
 * Une mesure déclenchée sur l'appareil vaut aussi « touche » (télécommande).
 */
@SuppressLint("MissingPermission")
@Suppress("DEPRECATION")
class DistoBle(private val ctx: Context, private val address: String, events: EventBus) : DistoDriver(events) {
    override val name = "disto_ble"

    companion object {
        val SERVICE: UUID = UUID.fromString("3ab10100-f831-4395-b29d-570977d5bf94")
        val CHAR_DISTANCE: UUID = UUID.fromString("3ab10101-f831-4395-b29d-570977d5bf94")
        val CHAR_COMMAND: UUID = UUID.fromString("3ab10109-f831-4395-b29d-570977d5bf94")
        val CCCD: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
    }

    private var gatt: BluetoothGatt? = null
    @Volatile private var cmdChar: BluetoothGattCharacteristic? = null
    @Volatile private var ready = CountDownLatch(1)
    @Volatile private var measureLatch: CountDownLatch? = null

    init { model = "Leica DISTO" }

    override fun connect() {
        val adapter = BluetoothHelper.adapter(ctx) ?: throw RuntimeException("Bluetooth indisponible")
        var dev: BluetoothDevice? = if (address.isNotBlank()) adapter.getRemoteDevice(address.uppercase()) else null
        if (dev == null) dev = BluetoothHelper.paired(ctx).firstOrNull { BluetoothHelper.nameOf(it).uppercase().contains("DISTO") }
        if (dev == null) dev = scanForDisto(adapter.bluetoothLeScanner)
        if (dev == null) throw RuntimeException("aucun DISTO trouvé (allumer le Bluetooth du DISTO)")
        model = BluetoothHelper.nameOf(dev)
        AppLog.d("DISTO : connexion à $model")
        ready = CountDownLatch(1)
        cmdChar = null
        gatt = dev.connectGatt(ctx, false, callback, BluetoothDevice.TRANSPORT_LE)
        if (!ready.await(30, TimeUnit.SECONDS)) { disconnect(); throw RuntimeException("connexion DISTO impossible (délai)") }
        if (cmdChar == null) { disconnect(); throw RuntimeException("service DISTO introuvable sur $model") }
        connected = true
        lastError = ""
        AppLog.d("DISTO : prêt")
    }

    private fun scanForDisto(scanner: BluetoothLeScanner?): BluetoothDevice? {
        if (scanner == null) return null
        var found: BluetoothDevice? = null
        val latch = CountDownLatch(1)
        val cb = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                val n = result.device.name ?: result.scanRecord?.deviceName ?: return
                if (n.uppercase().contains("DISTO")) { found = result.device; latch.countDown() }
            }
        }
        try {
            scanner.startScan(cb)
            latch.await(6, TimeUnit.SECONDS)
        } finally {
            try { scanner.stopScan(cb) } catch (ignored: Exception) {}
        }
        return found
    }

    private val callback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                g.discoverServices()
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                if (connected) {
                    connected = false
                    lastError = "DISTO déconnecté"
                    events.push("disto_disconnected")
                }
                ready.countDown()
            }
        }

        override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
            val svc = g.getService(SERVICE)
            if (svc != null) {
                val dist = svc.getCharacteristic(CHAR_DISTANCE)
                cmdChar = svc.getCharacteristic(CHAR_COMMAND)
                if (dist != null) {
                    g.setCharacteristicNotification(dist, true)
                    val cccd = dist.getDescriptor(CCCD)
                    if (cccd != null) {
                        cccd.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                        g.writeDescriptor(cccd)
                    }
                }
            }
            ready.countDown()
        }

        override fun onCharacteristicChanged(g: BluetoothGatt, c: BluetoothGattCharacteristic) {
            if (c.uuid != CHAR_DISTANCE) return
            val v = c.value ?: return
            if (v.size < 4) return
            val d = ByteBuffer.wrap(v, 0, 4).order(ByteOrder.LITTLE_ENDIAN).float.toDouble()
            onDistance(d)
        }
    }

    private fun onDistance(d: Double) {
        lastDistance = Math.round(d * 10000.0) / 10000.0
        lastKey = "measure"
        events.push("disto_measure", JSONObject().put("distance", lastDistance))
        events.push("disto_key", JSONObject().put("key", "measure"))
        measureLatch?.countDown()
    }

    override fun measure(sim: JSONObject?): Double {
        val g = gatt ?: throw RuntimeException("DISTO non connecté")
        val c = cmdChar ?: throw RuntimeException("DISTO non connecté")
        val latch = CountDownLatch(1)
        measureLatch = latch
        c.value = "g".toByteArray(Charsets.US_ASCII)
        if (!g.writeCharacteristic(c)) throw RuntimeException("commande DISTO refusée")
        if (!latch.await(10, TimeUnit.SECONDS)) throw RuntimeException("pas de mesure DISTO reçue")
        return lastDistance ?: throw RuntimeException("pas de mesure DISTO reçue")
    }

    override fun disconnect() {
        connected = false
        try {
            gatt?.disconnect()
            gatt?.close()
        } catch (ignored: Exception) {
        }
        gatt = null
        cmdChar = null
    }
}

File: bridge/drivers/disto_ble.py
# -*- coding: utf-8 -*-
"""
Pilote Leica DISTO Bluetooth Smart (D110 / D2 / D510 / D810...) via BLE (bleak).

Service DISTO  : 3ab10100-f831-4395-b29d-570977d5bf94
  distance     : 3ab10101-...  (float32 little endian, mètres, notification)
  commande     : 3ab10109-...  ("g" = déclencher une mesure, "o" = éteindre)
Les mesures déclenchées sur l'appareil arrivent par notification = usage
« télécommande » du logiciel de référence (une mesure disto = validation / distance).

Non testé sans appareil ; les UUID proviennent de la documentation Leica.
"""
import asyncio
import struct
import threading

from .base import DistoDriver

SERVICE = "3ab10100-f831-4395-b29d-570977d5bf94"
CHAR_DISTANCE = "3ab10101-f831-4395-b29d-570977d5bf94"
CHAR_COMMAND = "3ab10109-f831-4395-b29d-570977d5bf94"


class DistoBLE(DistoDriver):
    name = "disto_ble"

    def __init__(self, address, events):
        super().__init__(events)
        self.address = address
        self.loop = None
        self.client = None
        self.thread = None
        self._measure_event = None
        self.model = "Leica DISTO"

    def connect(self):
        from bleak import BleakClient, BleakScanner  # import tardif

        self.loop = asyncio.new_event_loop()
        self.thread = threading.Thread(target=self.loop.run_forever, daemon=True)
        self.thread.start()

        async def _connect():
            address = self.address
            if not address:
                devices = await BleakScanner.discover(timeout=6.0)
                for d in devices:
                    if d.name and "DISTO" in d.name.upper():
                        address = d.address
                        self.model = d.name
                        break
            if not address:
                raise RuntimeError("Aucun DISTO trouvé en BLE")
            self.client = BleakClient(address)
            await self.client.connect()
            await self.client.start_notify(CHAR_DISTANCE, self._on_distance)

        fut = asyncio.run_coroutine_threadsafe(_connect(), self.loop)
        fut.result(timeout=30)
        self.connected = True

    def _on_distance(self, _handle, data):
        try:
            d = struct.unpack("<f", bytes(data[:4]))[0]
        except Exception:
            return
        self.last_distance = round(float(d), 4)
        self.last_key = "measure"
        # une mesure déclenchée sur l'appareil vaut aussi "touche" (télécommande)
        self.events.push("disto_measure", {"distance": self.last_distance})
        self.events.push("disto_key", {"key": "measure"})
        if self._measure_event:
            self._measure_event.set()

    def measure(self, sim=None):
        if not self.client:
            raise RuntimeError("DISTO non connecté")
        self._measure_event = threading.Event()

        async def _trigger():
            await self.client.write_gatt_char(CHAR_COMMAND, b"g")

        asyncio.run_coroutine_threadsafe(_trigger(), self.loop).result(timeout=5)
        if not self._measure_event.wait(10):
            raise RuntimeError("Pas de mesure DISTO reçue")
        return self.last_distance

    def disconnect(self):
        self.connected = False
        if self.client and self.loop:
            async def _dc():
                try:
                    await self.client.disconnect()
                except Exception:
                    pass
            try:
                asyncio.run_coroutine_threadsafe(_dc(), self.loop).result(timeout=5)
            except Exception:
                pass
        if self.loop:
            self.loop.call_soon_threadsafe(self.loop.stop)
